cpp/ql/src/experimental/Security/CWE/CWE-783/OperatorPrecedenceLogicErrorWhenUseBitwiseOrLogicalOperations.ql

```diff
@@ -0,0 +1,179 @@
+/**
+ * @name Operator Precedence Logic Error When Use Bitwise Or Logical Operations
+ * @description --Finding places to use bit and logical operations, without explicit priority allocation.
+ *              --For example, `a || b ^ c` and `(a || b) ^ c` give different results when `b` is zero.
+ * @kind problem
+ * @id cpp/operator-precedence-logic-error-when-use-bitwise-logical-operations
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-783
+ *       external/cwe/cwe-480
```

Comment:
I would ask you to take your time with this fix.
I am still looking for false positives. and thanks to you, I hope to improve the speed of this search (pragma [inline]).
in my opinion this request has the potential for security.
as one of the proofs I can give an example from PR.

```cpp
#include <stdio.h>
#include <stdbool.h>
 
int main(void)
{
    bool a=1,b=0,c=1,res;

    res = a||b^c;
    printf("a||b^c:%i\n",res); // res==1
    res = a||(b^c);
    printf("a||(b^c):%i\n",res); // res==1
    
    res = (a||b)^c;
    printf("(a||b)^c:%i\n",res); // res==0

    return 0;
}
```

such errors can affect the progress of the program, as well as stay in the code for a long time.

Reply:
Readability and maintainability are both important for security, and I think we should keep the two (very relevant) CWE tags on the query. But I think the query has potential outside of just security, which is why I think the query deserves the `readability` and `maintainability` tags instead of `security`.

Reply:
thanks for clarifying your position.
if I understand correctly, then you think that the query does not find real errors, but finds places that can be stylistically improved. this is true, for example, for expressions that do not change the result (for example, `a||b&&c` or `a&&b&c`), but there are expressions that change the result depending on priority (for example, `a||b^c`) that can create an error in the code.

Reply:
Based on the two PRs you submitted based on the finding of this query (i.e., [difficulty in understanding priorities.](https://github.com/rui314/chibicc/pull/54) and [difficulty in understanding priorities.](https://github.com/Andarix/simutrans/pull/2)), it's my opinion that the query is mostly finding expressions with possibly confusing precedence/priority.

If you find a true-positive result where an incorrect precedence/priority causes a security-relevant issue, then I'd be happy to reconsider the tags.